File: Segundo/Monopoly/civitas_java (1)/proyecto_civitas/src/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import GUI.Dado;
public class Jugador implements Comparable<Jugador> {
    
    protected static int CasasMax=4;
    protected static int CasasPorHotel=4;
    protected boolean encarcelado;
    protected static int HotelesMax=4;
    private String nombre;
    private int numCasillaActual;
    protected static float PasoPorSalida=1000;
    protected static float PrecioLibertad=200;
    private boolean puedeComprar;
    protected float saldo;
    static float SaldoInicial=7500;
    protected ArrayList<TituloPropiedad> propiedades;
    protected SorpresaSalirCarcel salvoconducto;
    protected boolean especulador;
    

    public Jugador(String nom){
        nombre=nom;
        encarcelado=false;
        numCasillaActual=0;
        puedeComprar=true;
        saldo=7500;
        propiedades=new ArrayList<>();
        salvoconducto=null;
        especulador=false;    
    }
    protected Jugador(Jugador otro){
        this.nombre=otro.nombre;
        this.encarcelado=otro.encarcelado;
        this.numCasillaActual=otro.numCasillaActual;
        this.puedeComprar=otro.puedeComprar;
        this.saldo=otro.saldo;
        this.salvoconducto=otro.salvoconducto;
        this.propiedades=otro.propiedades;
        this.especulador=otro.especulador;
    }
    public boolean getEspeculador()
    {
        return especulador;
    }
    @Override
    public String toString() {
        String aux= "\nJugador: ";
        aux+= "\nNombre= " + nombre ;
        aux+= "\nEncarcelado= " + encarcelado ;
        aux+= "\nNumCasillaActual= " + numCasillaActual;
        aux+= "\nPuedeComprar= " + puedeComprar ;
        aux+= "\nSaldo= " + saldo;
        aux+= "\nPropiedades= " + propiedades.toString();
        aux+= "\nSalvoconducto= " + salvoconducto +"\n";
        return aux;
    }
    @Override
    public int compareTo(Jugador otro){

        return Float.compare(this.saldo, otro.saldo);
    }
    boolean comprar(TituloPropiedad titulo){
       boolean result=false;
       if(encarcelado)
           return result;
       if(puedeComprar){
           float precio=titulo.getPrecioCompra();
           if(puedoGastar(precio)){
              result=titulo.comprar(this);
              if(result){
                  propiedades.add(titulo);
                  Diario.getInstance().ocurreEvento("El jugador "+nombre+"compra la propiedad "+titulo.getNombre());
              }
              puedeComprar=false;
           }
       }
       
       return result;
       
    }
    boolean construirHotel(int ip){
        boolean result=false;
        if(encarcelado)
            return result;
       
        if(this.existeLaPropiedad(ip))
        {
            TituloPropiedad propiedad=propiedades.get(ip);
            boolean puedoEdificarHotel=this.puedoEdificarHotel(propiedad);
            float precio=propiedad.getPrecioEdificar();
            if(puedoGastar(precio) && propiedad.getNumHoteles()<this.getHotelesMax() && propiedad.getNumCasas()>=this.getCasasPorHotel()){
               puedoEdificarHotel=true;
            }
            if(puedoEdificarHotel)
            {
               result=propiedad.construirHotel(this);
               propiedad.derruirCasas(CasasPorHotel, this);
               Diario.getInstance().ocurreEvento("El jugador "+nombre+" construye hotel en la propiedad "+propiedades.get(ip).getNombre());
           }
        }

        return result; 
    }
    boolean construirCasa(int ip){
       boolean result=false;
       boolean puedoEdificarCasa=false;
       
        if(encarcelado)
           return result;
        else
        {
            if(existeLaPropiedad(ip)){
                TituloPropiedad propiedad=propiedades.get(ip);
                if(puedoEdificarCasa(propiedad))
                {
                    result=propiedad.construirCasa(this);
                    if(result)
                        Diario.getInstance().ocurreEvento("El jugador "+nombre+" construye casa en la propiedad "+propiedades.get(ip).getNombre());
                }
              
            }
        }
        return result;
       
    }
    
    public String getNombre() {
        return nombre;
    }
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    private void perderSalvoconducto (){
        salvoconducto.usada();
        salvoconducto=null;
    }
    int cantidadCasasHoteles(){
        int contador=0;
        for(int i=0; i<propiedades.size();i++)
        {
            contador+=propiedades.get(i).cantidadCasasHoteles();
        }
       return contador;
   }
    public float getSaldo()
    {
        return saldo;
    }
    protected Boolean existeLaPropiedad(int ip){
       return ip<=propiedades.size();
   }
    boolean tieneSalvoconducto(){
       return salvoconducto!=null;
    }
    boolean obtenerSalvoconducto(SorpresaSalirCarcel s){
       if(encarcelado)
           return false;
       else{
           salvoconducto=s;
           return true;     
       }
    }
    boolean hipotecar(int ip){
       boolean result=false;
       if(encarcelado)
           return result;
       
       if(existeLaPropiedad(ip)){
           TituloPropiedad propiedad=propiedades.get(ip);
           result=propiedad.hipotecar(this);
           if(result)
            Diario.getInstance().ocurreEvento("El jugador "+nombre+" hipoteca la propiedad "+propiedades.get(ip).getNombre());
       }

       return result;
   }
    public boolean isEncarcelado() {
        return encarcelado;
    }
    protected boolean debeSerEncarcelado(){
        boolean estado=false;
        if(isEncarcelado())
            estado=false;
        else
            if(!tieneSalvoconducto())
                estado=true;
            else
            {
                perderSalvoconducto();
                estado=false;
                Diario.getInstance().ocurreEvento("El jugador "+nombre +" se libra de la cárcel"); 
            }
        return estado;
           
   }
    boolean modificarSaldo(float cantidad){
        
        saldo+=cantidad;
        Diario.getInstance().ocurreEvento("Se ha modificado el sueldo del jugador "+nombre);
        return true;
    }
    

    boolean pagaAlquiler(float cantidad) {
        if(encarcelado)
           return false;
        else
           return paga(cantidad);
    }

    boolean paga(float cantidad) {
        return modificarSaldo(cantidad*-1);
    }
    boolean recibe(float cantidad){
        if(encarcelado)
            return false;
        else 
        {
            return modificarSaldo(cantidad); 
        }
    }
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        if(encarcelado)
            return false;
   
        if(propiedad.getNumCasas()<=CasasMax)
        return saldo>=propiedad.getPrecioEdificar();
        else
            return false;
          
    }
    
   
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        if(encarcelado)
            return false;
        else
        {
            if(propiedad.getNumHoteles()<=HotelesMax && propiedad.getNumCasas()==CasasMax)
                return saldo>=propiedad.getPrecioEdificar();
            else
                return false;      
        }
           
    }

    boolean puedeComprarCasilla() {
        puedeComprar= !encarcelado;
        return puedeComprar;
    }
    
    boolean vender(int ip)
    {
        boolean estado = false;
        if ((!encarcelado)&&(existeLaPropiedad(ip)))
        {
            estado=propiedades.get(ip).vender(this);
            if (estado)
            {
                String msg="La propiedad "+ propiedades.get(ip).getNombre() +" ha sido vendida.";
                Diario.getInstance().ocurreEvento(msg);
                propiedades.remove(ip);
            }
        }
        return estado;
    }
    
    
    boolean pagaImpuesto(float cantidad) {
        if(encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    protected boolean puedoGastar (float precio){
       if(encarcelado)
            return false;
       else
            return saldo>=precio;
   }
    boolean encarcelar(int numCasillaCarcel) {
        if(debeSerEncarcelado())
        {
            encarcelado=true;
            moverACasilla(numCasillaCarcel);
            Diario.getInstance().ocurreEvento("Jugador encarcelado: "+nombre);
        } 
        return encarcelado;
    }

    boolean tieneAlgoQueGestionar() {
        return propiedades!=null;
    }
    boolean moverACasilla(int numCasilla){
        boolean estado=false;
        if(!encarcelado){
           numCasillaActual=numCasilla;
           puedeComprar=false; 
           Diario.getInstance().ocurreEvento("El jugador "+ nombre + " desplazado a: "+ Float.toString(numCasillaActual));
           estado= true;
        }
        return estado;
   }
    
    private Boolean puedeSalirCarcelPagando(){
        return saldo>=PrecioLibertad;
    }
    boolean pasaPorSalida(){
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento("El jugador "+nombre+" pasa por la casilla Salida");
        return true;
    }
    boolean salirCarcelPagando(){
        boolean estado=false;
        if(encarcelado && puedeSalirCarcelPagando()){
           paga(PrecioLibertad);
           encarcelado=false;
           Diario.getInstance().ocurreEvento("El jugador "+nombre+" ha salido de la carcel pagando");
           estado= true;
        }
       
        return estado;
    }
    boolean salirCarcelTirando(){
        if(Dado.getInstance().salgoDeLaCarcel()){
            encarcelado=false;
            Diario.getInstance().ocurreEvento("El jugador "+nombre+" ha salido de la carcel tirando");
            return true;
        }
        else
            return false;
    }
    boolean enBancarrota(){
       return saldo<=0;
    }
    protected int getCasasMax(){
       return CasasMax;
    }
   
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
   
    protected int getHotelesMax(){
        return HotelesMax;
    }
    public ArrayList<String> getNombrePropiedades(){
       ArrayList<String> aux = new ArrayList<>();
       
       for(int i=0;i<propiedades.size();i++)
           aux.add(propiedades.get(i).getNombre());
       
       return aux;
       
    }
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
   
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }
    public ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
   
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    boolean cancelarHipoteca(int ip){
        boolean result=false;
        if(encarcelado)
            return result;
       
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad=propiedades.get(ip);
            float cantidad=propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar=puedoGastar(cantidad);
            
            if(puedoGastar){
                result=propiedad.cancelarHipoteca(this);

                if(result)
                  Diario.getInstance().ocurreEvento("El jugador "+nombre+" cancela la hipoteca de la propiedad "+propiedades.get(ip).getNombre());
                
            }
        }
        
        return result;
       
       
   }
}

File: Segundo/Monopoly/civitas_ruby (1)/proyecto_civitas_ruby/pkg/proyecto_civitas_ruby-0.0.1/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


module Civitas
  class Jugador
    include Comparable
    attr_reader :nombre,:numCasillaActual,:HotelesMax,:CasasMax,:CasasPorHotel,:propiedades,:puedeComprar,:saldo,:encarcelado,:valor,:salvoconducto,:especulador
    

    def initialize(nom)
      @nombre=nom
      @encarcelado=false
      @numCasillaActual=0
      @puedeComprar=true
      @saldo=7500
      @propiedades=Array.new
      @salvoconducto=nil
      @@CasasMax=4
      @@CasasPorHotel=4
      @@HotelesMax=4
      @@PasoPorSalida=1000
      @@PrecioLibertad=200
      @@SaldoInicial=7500
      @especulador = false
    end
    def jugador_copia(otro)
      @nombre = otro.nombre
      @encarcelado=otro.encarcelado
      @numCasillaActual=otro.numCasillaActual
      @puedeComprar=otro.puedeComprar
      @saldo=otro.saldo
      @salvoconducto= otro.salvoconducto
      @propiedades=otro.propiedades
      for i in @propiedades
        i.actualiza_propietario_por_conversion(self)  
      end
      @especulador = otro.especulador
  end
    def self.copia(otro)
      @nombre = otro.getNombre
      @encarcelado=otro.isEncarcelado
      @numCasillaActual=otro.getNumCasillaActual
      @puedeComprar=otro.getPuedeComprar
      @saldo=otro.getSaldo
      @salvoconducto= otro.salvoconducto
      @propiedades=otro.propiedades   
    end

    def from(p)

      init(p.salvoconducto,p.isEncarcelado,p.nombre,p.numCasillaActual,p.saldo,p.puedeComprar)
      @propiedades = p.propiedades
    end

    def init(salvo,carcel,nombre,casilla,saldo,compra)

      @salvoconducto=[]
      isEncarcelado=carcel
      @nombre=nombre
      @numCasillaActual=casilla
      @saldo=saldo
      @puedeComprar=compra
    end

    

    def cancelarHipoteca(ip)
      result = false
      if (@encarcelado)
        return result
      end
      if (existeLaPropiedad(ip))
        propiedad = @propiedades.at(ip)
        cantidad = propiedad.getImporteCancelarHipoteca
        puedoGastar= puedoGastar(cantidad);
        if (puedoGastar)
          result = propiedad.cancelarHipoteca(self)
          if result
            Diario.instance.ocurre_evento("El jugador "+@nombre+" cancela la hipoteca de la propiedad "+ip.to_s)
          end

        end
      end
      return result


    end



    def cantidadCasasHoteles() 
      cont=0
      for i in 0...@propiedades.size
        cont=cont+@propiedades[i].cantidad_casas_hoteles
      end
      return cont
    end
    def <=>(otro)
    @valor<=>otro.valor
    end
#    def compareTo(otro)#1 if self>other; 0 if self==other; -1 if self<other
#      
#      
#      if (@saldo > otro.getSaldo)
#        valor=1
#      
#      else
#        if @saldo == otro.getSaldo()
#          valor=0
#        else
#          valor=-1
#        end
#      end
#      
#      @valor <=> otro.getSaldo)
#      return valor
#    end

    def comprar(titulo ) 
      result = false
      if (@encarcelado)
        return result
      end

      if (@puedeComprar)
        precio = titulo.precioCompra
        if (puedoGastar(precio))
          result = titulo.comprar(self)
          if (result)
            @propiedades.push(titulo)
            Diario.instance.ocurre_evento("El jugador "+@nombre+ " compra la propiedad "+titulo.toString())
          end
          @puedeComprar=false
        end

      end

      return result
    end

    def construirCasa(ip ) 
      result = false
      if @encarcelado
        return result
      else
        if existeLaPropiedad(ip)

        propiedad= @propiedades.at(ip)
       
          if (puedoEdificarCasa(propiedad))
            result=propiedad.construirCasa(self)
            if (result)
               Diario.instance.ocurre_evento("El jugador "+@nombre+ " construye casa en la propiedad "+ip.to_s)
            end
          end
        end
      end
    end

    def construirHotel(ip ) 
      result = false
      if @encarcelado
        return result
      end
      if existeLaPropiedad(ip)

        propiedad= @propiedades.at(ip)
        puedoEdificarHotel = puedoEdificarHotel(propiedad)
        precio=propiedad.precioEdificar
        if puedoEdificarHotel

        result= propiedad.construirHotel(self)
        propiedad.derruirCasas( @@CasasPorHotel, self)

        Diario.instance.ocurre_evento("El jugador "+@nombre+ " construye hotel en la propiedad "+ip.to_s)



        end
      end
      return result
    end

    def debeSerEncarcelado() 
      estado=false
      if (@encarcelado)
        estado=false

      else
        if tieneSalvoconducto()==false
          estado=true
        else
          perderSalvoConducto
          estado=false
          Diario.instance.ocurre_evento("El jugador se libra de la carcel")
        end
      end





      return estado
    end

    def enBancarrota() 

      estado=true
      if @saldo >=0
        estado=false
      end
      return estado

    end

    def encarcelar(numCasillaCarcel) 


      if debeSerEncarcelado
        @encarcelado=true
        moverACasilla(numCasillaCarcel)
        Diario.instance.ocurre_evento("Jugador encarcelado")

      end


      return @encarcelado
    end

    private
    def existeLaPropiedad(ip)
      return ip<=@propiedades.size
    end
    public
    def getCasasMax()
      @@CasasMax
    end
    private
    def getCasasPorHotel()
      @@CasasPorHotel
    end
    public
    def getHotelesMax()
      return @@HotelesMax
    end

    public 
    def getNombre()
      return @nombre
    end 
    

    def getNumCasillaActual()
      return @numCasillaActual
    end
    private 
    def getPrecioLibertad() 
      return @@PrecioLibertad
    end

    def getPremioPasoSalida() 
      return @@PasoPorSalida
    end

    public
    def getPropiedades() #: TituloPropiedad[1..*]
      return @propiedades
    end
    def get_nombre_propiedades
      nombres=Array.new
      for i in 0...@propiedades.size
        nombres<<@propiedades[i].nombre
      end
      return nombres
    end

    def getPuedeComprar() 
      return @puedeComprar
    end

    def getSaldo() 
      return @saldo
    end

    def hipotecar(ip )
     result = false
     if (@encarcelado)
       return result
     end
     if (existeLaPropiedad(ip))
        propiedad = @propiedades.at(ip)
        result = propiedad.hipotecar(self)
        if result
          Diario.instance.ocurre_evento("El jugador "+@nombre+ " hipoteca la propiedad "+ip.to_s)
        end
     end

     return result
    end 

    def isEncarcelado() 
      return @encarcelado
    end

    def modificarSaldo(cantidad )
      estado=true

      @saldo=@saldo+cantidad
      Diario.instance.ocurre_evento('Se ha modificado el saldo del jugador')
      return estado
    end

    
    def moverACasilla(numCasilla ) 
      estado=false
      if !@encarcelado
        @numCasillaActual=numCasilla
        @puedeComprar=false
        msg=@nombre + " desplazado a #{numCasilla}"
        Diario.instance.ocurre_evento(msg)
        estado=true
      end
      return estado
    end

    def obtenerSalvoconducto(sorpresa) 
      estado = false
        if (@encarcelado==false)
          @salvoconducto=sorpresa
          estado=true
        end

      return estado
    end
    def paga(cantidad ) 
      return estado=modificarSaldo(cantidad * (-1))
    end

    def pagaAlquiler(cantidad) 
      estado=false
      if @encarcelado
        estado=false
      else
        estado=paga(cantidad)
      end
      return estado
    end

    def pagaImpuesto(cantidad )
      estado=false
      if @encarcelado
        estado=false
      else
        estado=paga(cantidad)
      end
      return estado
    end

    def pasaPorSalida() 
      estado=true
      modificarSaldo(@@PasoPorSalida)
 
      Diario.instance.ocurre_evento("El jugador ha pasado por la salida")
      return estado
    end
    private
    def perderSalvoConducto() 
      @salvoconducto.usada
      @salvoconducto=nil
    end
    public
    def puedeComprarCasilla() 
      if(@encarcelado)
        @puedeComprar=false
      else
        @puedeComprar=true
      end
      return @puedeComprar
    end
    private
    def puedeSalirCarcelPagando() 
      estado = false
      if @saldo >= @@PrecioLibertad
        estado=true
      end
      return estado
    end
    def puedoEdificarCasa(propiedad ) 
      estado=false
     
        if((propiedad.numCasas<=@@CasasMax)&&!@encarcelado)
          if puedoGastar(propiedad.precioEdificar ) 
            estado=true
          end
        end

     
      return estado
    end
    def puedoEdificarHotel(propiedad ) 
      if(@encarcelado)
        return false
      else
        
      if(propiedad.numHoteles<=@@HotelesMax && propiedad.numCasas==4)
        return @saldo>=propiedad.precioEdificar
      else
        return false
      end
    end
      end


    def puedoGastar(precio ) 
      estado=false
      if @encarcelado==false
        if(@saldo >= precio)
          estado=true
        end
      end
      return estado
    end
    public
    def recibe(cantidad ) 
      estado=false
      if @encarcelado
        estado=false
      else
        estado=modificarSaldo(cantidad)
      end
      return estado
    end
    def salirCarcelPagando() 
      estado=false
        if (@encarcelado)
          if puedeSalirCarcelPagando
            paga(@@PrecioLibertad)
            estado=true
            @encarcelado=false
            msg=@nombre +" : El jugador ha pagado por salir de la carcel"
            Diario.instance.ocurre_evento(msg)
          end
        end
      return estado
    end
    def salirCarcelTirando() 
      estado=false
      if Dado.instance.salgoDeLaCarcel
        @encarcelado=false
        Diario.instance.ocurre_evento("El jugador ha salido de la carcel con el dado")
        estado=true
      end

      return estado
    end
    def tieneAlgoQueGestionar() 
      return @propiedades!=nil
    end
    def tieneSalvoconducto() 
      estado=false
        if @salvoconducto!=nil
          estado=true
        end
      return estado
    end
    def toString() 
      aux="\nJugador:
          nombre:"+ @nombre+"
          encarcelado: "+@encarcelado.to_s+"
          saldo: "+@saldo.to_s+"
          casilla actual:"+ @numcasillaActual.to_s+"\n"
      aux2="Propiedades= "
      for i in 0...@propiedades.size
        aux2=aux2+@propiedades[i].nombre
      end
      return aux+aux2
    end
    def vender(ip ) 
      estado=false
      if @encarcelado ==false
        if existeLaPropiedad(ip) 
          estado=@propiedades[ip].vender(self) #self.instance
          if (estado)
           Diario.instance.ocurre_evento("La propiedad "+@propiedades[ip].nombre+" ha sido vendida por el jugador #{@nombre}")
           @propiedades.delete_at(ip) 
          end
        end

      end
      return estado
    end 
  end
end

